net/graphsync_fetcher_test.go

```diff
@@ -533,6 +536,180 @@ func TestGraphsyncFetcher(t *testing.T) {
 		require.Error(t, err, "invalid receipts for for receipt collection (cid %s)", final.At(0).MessageReceipts.String())
 	})
 
+	t.Run("hangup occurs during first layer fetch but recovers through fallback", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		defer cancel()
+		gen := builder.NewGenesis()
+		final := builder.BuildOn(gen, 3, withMessageEachBuilder)
+		height, err := final.Height()
+		require.NoError(t, err)
+		chain0 := types.NewChainInfo(pid0, final.Key(), height)
+		chain1 := types.NewChainInfo(pid1, final.Key(), height)
+		chain2 := types.NewChainInfo(pid2, final.Key(), height)
+		pt := newFakePeerTracker(chain0, chain1, chain2)
+
+		mgs := newMockableGraphsync(ctx, bs, t)
+		mgs.expectRequestToRespondWithLoader(pid0, layer1Selector, loader, final.At(0).Cid())
+		mgs.expectRequestToRespondWithHangupAfter(pid0, layer1Selector, loader, 0, final.At(1).Cid(), final.At(2).Cid())
+		mgs.expectRequestToRespondWithLoader(pid1, layer1Selector, loader, final.At(1).Cid())
+		mgs.expectRequestToRespondWithHangupAfter(pid1, layer1Selector, loader, 0, final.At(2).Cid())
+		mgs.expectRequestToRespondWithLoader(pid2, layer1Selector, loader, final.At(2).Cid())
+		mgs.expectRequestToRespondWithLoader(pid2, recursiveSelector(1), loader, final.At(0).Cid())
+
+		fetcher := net.NewGraphSyncFetcher(ctx, mgs, bs, bv, pt, net.UseUnresponsiveTimeout(100*time.Millisecond))
```

Comment:
Blocking: any test that uses `time` is 1) automatically a `tf.IntegrationTest(t)` because it's slow, and 2) probably nondeterministic and flaky.

It may be ok to add some integration tests (in a new file) with very long timeouts to hope for non-flakiness, but in this case we can probably do much better by explicitly controlling time.

Reply:
I'm just gonna factor out time somehow :) No worries. Nice that we have this design decision already :)